lib/Doctrine/Persistence/ObjectRepository.php

```diff
@@ -16,14 +18,14 @@ interface ObjectRepository
      *
      * @return object|null The object.
      */
-    public function find($id);
+    public function find($id) : ?object;
 
     /**
      * Finds all objects in the repository.
      *
      * @return object[] The objects.
      */
-    public function findAll();
+    public function findAll() : array;
```

Comment:
I believe we should typehint `iterable` here. For example, passing an array means hydrating every object directly after the query, while we may be able to return a lazy iterator otherwise (as is done in MongoDB ODM in some instances).
```suggestion
    public function findAll() : iterable;
```

Note: this also applies to other methods that return an `array` in this interface.

Reply:
That's too big BC break. Userspace needs to be able to keep relying on possibility to use array_ functions here

Reply:
Ah, usage of array functions was something I hadn't considered. However, I still think it's worth it to allow returning things other than arrays.

Reply:
We should do this work in another specific PR. Adding new methods instead of changing existing ones would probably be better imo too.